### ISSUE TITLE
[Outreachy 2023] Applicants tasks/contributions

### ISSUE THREAD

**Issue by bluca** [systemd/systemd#26585]
# Intro

This ticket will serve as the central point tracking the contributions of perspective Outreachy (spring 2023 cohort) applicants.

For those interested, this process is described here: https://www.outreachy.org/apply/make-contributions/
The systemd project will participate again this year: https://www.outreachy.org/communities/cfp/systemd/
https://www.outreachy.org/outreachy-may-2023-internship-cohort/communities/systemd/#improve-user-experience-on-early-system-boot-failu

A series of tasks will be listed here. Some of them will be "repeatable", ie: some basic tasks that every applicant can do. Some can only be done once, so they are first-come first-served. Some will be relatively quick and should be doable in a couple hours, others might take several days, and they'll be tagged as such.

Every applicant should start with the basic repeatable tasks, and complete at least two. Simply quote the task in a comment, posting proof of completion. We will "thumbs up" the comment to acknowledge completion.

After at least two of those are done and acked, then applicants should move on to the "non-repeatable" tasks. Write a comment quoting the task you selected, and if it's open-ended specify what exactly you are going to do, before you start. Those should end up with a pull request. Note that simply opening a pull request is not enough, they will be reviewed and might need follow-ups before they can be merged, so take that into account when planning for time.

Applicants should complete and wait for ack on completion at least 2 of the repeatable tasks before moving on to the short/long tasks.

# Documentation for new contributors
Contributors who are new to `systemd` are highly encouraged to start from the following documents:

- https://systemd.io/CONTRIBUTING/
- https://systemd.io/CODING_STYLE/
- https://systemd.io/HACKING/
- https://systemd.io/CODE_QUALITY/
- https://systemd.io/ARCHITECTURE/

# Repeatable tasks

- clone the project, build it, add an `Hello World from <name>!` log message to a tool of your choice from `src/`, run it and post a screenshot as proof
- clone the project, build it, pick one tool under `src/` of your choice and run it, inspect the code and in your own word write a paragraph explaining what the code (or a section of the code, some are quite large!) does - note that we don't mean just from the point of view of a user, as that's explained in the manual, but describe the actual code
- study the documentation, and write a paragraph summarizing one of the documents in your own words. Pick at least one medium/large size document, OR at least three smaller ones. For easier viewing, documents under `docs/` are rendered on https://systemd.io/ and documents under `man/` are rendered on https://www.freedesktop.org/software/systemd/man/

# Short tasks
- fix one of the issues labelled as "please-submit-as-pr" https://github.com/systemd/systemd/issues?q=is%3Aopen+is%3Aissue+label%3Aplease-submit-as-pr
- improve an existing unit test `src/test/` to increase code coverage for the basic common code - see under `src/basic/` on https://coveralls.io/github/systemd/systemd
- add a new unit test to cover the 'shared' common code - see under `src/shared/` on https://coveralls.io/github/systemd/systemd
- update outdated executables/daemons to bus_call_method/bus_get_property: https://github.com/systemd/systemd/issues/20928#issuecomment-933718629
- update outdated executables/daemons to new-style sd-bus macros: https://github.com/systemd/systemd/issues/20928#issuecomment-933733352

# Long tasks
Please note that the description might sound simple, but in reality the task might require more time than you have available. Before starting, state in a comment which issue/TODO you'd like to tackle and how much time you have available, and we'll try to advise on whether it's enough to complete it.

- fix one of the issues labelled as "RFE" https://github.com/systemd/systemd/issues?q=is%3Aopen+is%3Aissue+label%3A%22RFE+%F0%9F%8E%81%22
- fix one of the TODO items https://github.com/systemd/systemd/blob/main/TODO
- add an integration test `test/README.testsuite` for a tool/daemon that is not currently covered

**Comment by sidrahh12** [comment]
Hi, I have cloned the project, how do I
"add an Hello World from <name>! log message to a tool of your choice from src/" 
do this part of the task?

**Comment by 1awesomeJ** [comment]
Hello everyone, Josh here.
Please I'm having issues cloning the project. I keep getting the error:

 "error: invalid path 'test/fuzz/fuzz-unit-file/dev-mapper-fedora_krowka\x2dswap.swap'
fatal: unable to checkout working tree"

I tried cloning with git bash on a windows machine.

I will attempt now on an ubuntu instance.

Any help is appreciated. 

Thanks.

**Comment by bluca** [comment]
> Hi, I have cloned the project, how do I build it (I tried following the instructions in the README file) "add an Hello World from ! log message to a tool of your choice from src/" do this part of the task?

Please read the documents linked above, they explain how to build the project using meson

**Comment by bluca** [comment]
> Hello everyone, Josh here. Please I'm having issues cloning the project. I keep getting the error:
> 
> "error: invalid path 'test/fuzz/fuzz-unit-file/dev-mapper-fedora_krowka\x2dswap.swap' fatal: unable to checkout working tree"
> 
> I tried cloning with git bash on a windows machine.
> 
> I will attempt now on an ubuntu instance.
> 
> Any help is appreciated.
> 
> Thanks.

Windows is not supported, you need to use Linux to develop and run this project

**Comment by 1awesomeJ** [comment]
Thanks Bluca. It's 1:15am local rn. I'll resume to it in the morning. Thanks again.

**Comment by sidrahh12** [comment]
`root@Ahmed-Laptop:~# git clone https://github.com/systemd/systemd.git
fatal: destination path 'systemd' already exists and is not an empty directory.
root@Ahmed-Laptop:~# cd systemd
root@Ahmed-Laptop:~/systemd# mkdir build
mkdir: cannot create directory ‘build’: File exists
root@Ahmed-Laptop:~/systemd# cd build
root@Ahmed-Laptop:~/systemd/build# meson ..
The Meson build system
Version: 0.61.2
Source dir: /root/systemd
Build dir: /root/systemd/build
Build type: native build
Project name: systemd
Project version: 253
C compiler for the host machine: cc (gcc 11.3.0 "cc (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0")
C linker for the host machine: cc ld.bfd 2.38
Host machine cpu family: x86_64
Host machine cpu: x86_64
Program tools/add-git-hook.sh found: YES (/root/systemd/tools/add-git-hook.sh)
Program tools/meson-build.sh found: YES (/root/systemd/tools/meson-build.sh)
^[[BC++ compiler for the host machine: c++ (gcc 11.3.0 "c++ (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0")
C++ linker for the host machine: c++ ld.bfd 2.38
Compiler for C supports arguments -Wno-missing-field-initializers: YES
Compiler for C supports arguments -Wno-unused-parameter: YES
Compiler for C supports arguments -Warray-bounds: YES
Compiler for C supports arguments -Warray-bounds=2: YES
Compiler for C supports arguments -Wdate-time: YES
Compiler for C supports arguments -Wendif-labels: YES
Compiler for C supports arguments -Werror=format=2: YES
Compiler for C supports arguments -Werror=format-signedness: YES
Compiler for C supports arguments -Werror=implicit-function-declaration: YES
Compiler for C supports arguments -Werror=implicit-int: YES
Compiler for C supports arguments -Werror=incompatible-pointer-types: YES
Compiler for C supports arguments -Werror=int-conversion: YES
Compiler for C supports arguments -Werror=overflow: YES
Compiler for C supports arguments -Werror=override-init: YES
Compiler for C supports arguments -Werror=return-type: YES
Compiler for C supports arguments -Werror=shift-count-overflow: YES
Compiler for C supports arguments -Werror=shift-overflow=2: YES
Compiler for C supports arguments -Werror=undef: YES
Compiler for C supports arguments -Wfloat-equal: YES
Compiler for C supports arguments -Wimplicit-fallthrough=5: YES
Compiler for C supports arguments -Winit-self: YES
Compiler for C supports arguments -Wlogical-op: YES
Compiler for C supports arguments -Wmissing-include-dirs: YES
Compiler for C supports arguments -Wmissing-noreturn: YES
Compiler for C supports arguments -Wnested-externs: YES
Compiler for C supports arguments -Wold-style-definition: YES
Compiler for C supports arguments -Wpointer-arith: YES
Compiler for C supports arguments -Wredundant-decls: YES
Compiler for C supports arguments -Wshadow: YES
Compiler for C supports arguments -Wstrict-aliasing=2: YES
Compiler for C supports arguments -Wstrict-prototypes: YES
Compiler for C supports arguments -Wsuggest-attribute=noreturn: YES
Compiler for C supports arguments -Wunused-function: YES
Compiler for C supports arguments -Wwrite-strings: YES
Compiler for C supports arguments -Wzero-length-bounds: YES
Compiler for C supports arguments -Wno-error=#warnings: NO
Compiler for C supports arguments -Wno-string-plus-int: NO
Compiler for C supports arguments -fstrict-flex-arrays: NO
Compiler for C supports arguments -Wno-unused-result: YES
Compiler for C supports arguments -Werror=missing-declarations: YES
Compiler for C supports arguments -Werror=missing-prototypes: YES
Compiler for C supports arguments -fdiagnostics-show-option: YES
Compiler for C supports arguments -fno-common: YES
Compiler for C supports arguments -fno-strict-aliasing: YES
Compiler for C supports arguments -fstack-protector: YES
Compiler for C supports arguments -fstack-protector-strong: YES
Compiler for C supports arguments -fstrict-flex-arrays=1: NO
Compiler for C supports arguments -fvisibility=hidden: YES
Compiler for C supports arguments --param=ssp-buffer-size=4: YES
Compiler for C supports arguments -fno-omit-frame-pointer: YES
Compiler for C supports link arguments -Wl,--fatal-warnings: YES
Compiler for C supports link arguments -Wl,-z,now: YES
Compiler for C supports link arguments -Wl,-z,relro: YES
Compiler for C supports link arguments -fstack-protector: YES
Compiler for C supports link arguments -Wl,--warn-common: YES
Compiler for C supports arguments -Wzero-length-bounds: YES (cached)
Checking if "-Werror=shadow with local shadowing" compiles: YES
Compiler for C++ supports arguments -Wno-missing-field-initializers: YES
Compiler for C++ supports arguments -Wno-unused-parameter: YES
Compiler for C supports arguments -Wstringop-truncation: YES
Checking for size of "dev_t" : 8
Checking for size of "ino_t" : 8
Checking for size of "time_t" : 8
Checking for size of "rlim_t" : 8
Checking for size of "typeof(((struct timex *)0)->freq)" : 8
Computing int of "LONG_MAX" : 9223372036854775807
Checking for size of "char16_t" : 2
Checking for size of "char32_t" : 4
Checking for size of "struct mount_attr" : -1
Checking for size of "struct mount_attr" : 32
Checking for size of "struct statx" : 256
Checking for function "secure_getenv" : YES
Checking for function "__secure_getenv" : NO
Checking for function "memfd_create" : YES
Checking for function "gettid" : YES
Checking for function "pivot_root" : NO
Checking for function "ioprio_get" : NO
Checking for function "ioprio_set" : NO
Checking for function "name_to_handle_at" : YES
Checking for function "setns" : YES
Checking for function "renameat2" : YES
Checking for function "kcmp" : NO
Checking for function "keyctl" : NO
Checking for function "copy_file_range" : YES
Checking for function "bpf" : NO
Checking for function "statx" : YES
Checking for function "explicit_bzero" : YES
Checking for function "reallocarray" : YES
Checking for function "set_mempolicy" : NO
Checking for function "get_mempolicy" : NO
Checking for function "pidfd_send_signal" : NO
Checking for function "pidfd_open" : NO
Checking for function "rt_sigqueueinfo" : NO
Checking for function "rt_tgsigqueueinfo" : NO
Checking for function "mallinfo" : YES
Checking for function "mallinfo2" : YES
Checking for function "execveat" : YES
Checking for function "close_range" : YES
Checking for function "epoll_pwait2" : YES
Checking for function "mount_setattr" : NO
Checking for function "move_mount" : NO
Checking for function "open_tree" : NO
Checking for function "fsopen" : NO
Checking for function "fsconfig" : NO
Checking for function "getdents64" : YES
Checking for function "getrandom" : YES
Program sh found: YES (/usr/bin/sh)
Program echo found: YES (/usr/bin/echo)
Program sed found: YES (/usr/bin/sed)
Program awk found: YES (/usr/bin/awk)
Program stat found: YES (/usr/bin/stat)
Program ln found: YES (/usr/bin/ln)
Program git found: YES (/usr/bin/git)
Program env found: YES (/usr/bin/env)
Program perl found: YES (/usr/bin/perl)
Program rsync found: YES (/usr/bin/rsync)
Program quotaon /usr/sbin/quotaon /sbin/quotaon found: NO
Program quotacheck /usr/sbin/quotacheck /sbin/quotacheck found: NO
Program kmod found: YES (/usr/bin/kmod)
Program kexec /usr/sbin/kexec /sbin/kexec found: NO
Program sulogin found: YES (/usr/sbin/sulogin)
Program mount found: YES (/usr/bin/mount)
Program umount found: YES (/usr/bin/umount)
Program loadkeys found: YES (/usr/bin/loadkeys)
Program setfont found: YES (/usr/bin/setfont)
Program nologin found: YES (/usr/sbin/nologin)
Program gperf found: NO

../meson.build:726:0: ERROR: Program 'gperf' not found or not executable

A full log can be found at /root/systemd/build/meson-logs/meson-log.txt
root@Ahmed-Laptop:~/systemd/build# cd build
root@Ahmed-Laptop:~/systemd/build/build# ninja
ninja: error: loading 'build.ninja': No such file or directory
root@Ahmed-Laptop:~/systemd/build/build# sudo ninja install
ninja: error: loading 'build.ninja': No such file or directory
root@Ahmed-Laptop:~/systemd/build/build#`


**I am getting stuck here what should I do? I used Ubuntu.**

**Comment by bluca** [comment]
On Ubuntu you can use `apt build-dep systemd` to install the required build dependencies

**Comment by sidrahh12** [comment]
`root@Ahmed-Laptop:~/systemd/build/build# apt build-dep systemd
Reading package lists... Done
E: You must put some 'deb-src' URIs in your sources.list
root@Ahmed-Laptop:~/systemd/build/build#`

It shows this now, what do I do next?

**Comment by bluca** [comment]
Use your favourite search engine to figure out how to solve that error

**Comment by sidrahh12** [comment]
alrightt, thank you!

**Comment by sidrahh12** [comment]
> Use your favourite search engine to figure out how to solve that error

`root@Ahmed-Laptop:~/systemd/build# apt build-dep systemd
Reading package lists... Done
E: You must put some 'deb-src' URIs in your sources.list
root@Ahmed-Laptop:~/systemd/build# sudo nano /etc/apt/sources.list
root@Ahmed-Laptop:~/systemd/build# sudo apt-get update
Hit:1 http://archive.ubuntu.com/ubuntu jammy InRelease
Hit:2 http://security.ubuntu.com/ubuntu jammy-security InRelease
Hit:3 http://archive.ubuntu.com/ubuntu jammy-updates InRelease
Hit:4 http://archive.ubuntu.com/ubuntu jammy-backports InRelease
Get:5 http://archive.ubuntu.com/ubuntu bionic InRelease [242 kB]
Err:5 http://archive.ubuntu.com/ubuntu bionic InRelease
  The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 3B4FE6ACC0B21F32
Reading package lists... Done
W: GPG error: http://archive.ubuntu.com/ubuntu bionic InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 3B4FE6ACC0B21F32
E: The repository 'http://archive.ubuntu.com/ubuntu bionic InRelease' is not signed.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
root@Ahmed-Laptop:~/systemd/build# apt build-dep systemd
Reading package lists... Done
E: You must put some 'deb-src' URIs in your sources.list`

I seem to be stuck again sir 😅

**Comment by 1awesomeJ** [comment]
Good afternoon Bluca, I'm sorry if my questions sound naïve. I just completed the build with memo, and I've don "ninja install".
Looking at the src/ directory, the tool I'd like to log my message on is the "ac-power" tool.

1. I'm yet to figure out how to run it, is there a  documentation that helps? The links to the systemd.io pages above have docs on CONTRIBUTING, They don't seem to have on BUILDING and RUNNING.


2. I'm trying to run the "ac-power" tool on an aws ubuntu instance that i connect to via ssh with no GUI, would that be possible? If not, is there a documentation I can read to know which tools would work on my CLI?

THank you sir.

**Comment by bluca** [comment]
> Good afternoon Bluca, I'm sorry if my questions sound naïve. I just completed the build with memo, and I've don "ninja install". Looking at the src/ directory, the tool I'd like to log my message on is the "ac-power" tool.
> 
> 1. I'm yet to figure out how to run it, is there a  documentation that helps? The links to the systemd.io pages above have docs on CONTRIBUTING, They don't seem to have on BUILDING and RUNNING.

Just call the built binary `$ build/systemd-ac-power --help`

> 2. I'm trying to run the "ac-power" tool on an aws ubuntu instance that i connect to via ssh with no GUI, would that be possible? If not, is there a documentation I can read to know which tools would work on my CLI?

That should be fine

**Comment by bluca** [comment]
> I seem to be stuck again sir sweat_smile

You will need to use the internet and existing help forums outside from here to figure out and solve basic Linux usage steps. As outlined in the project proposal, minimal experience in running Linux as a user is needed.

**Comment by 1awesomeJ** [comment]
> > Good afternoon Bluca, I'm sorry if my questions sound naïve. I just completed the build with memo, and I've don "ninja install". Looking at the src/ directory, the tool I'd like to log my message on is the "ac-power" tool.
> > 
> > 1. I'm yet to figure out how to run it, is there a  documentation that helps? The links to the systemd.io pages above have docs on CONTRIBUTING, They don't seem to have on BUILDING and RUNNING.
> 
> Just call the built binary `$ build/systemd-ac-power --help`
> 
> > 2. I'm trying to run the "ac-power" tool on an aws ubuntu instance that i connect to via ssh with no GUI, would that be possible? If not, is there a documentation I can read to know which tools would work on my CLI?
> 
> That should be fine

Thank you Bluca. Your response is always prompt. Quite honoured to be here sir.

**Comment by 1awesomeJ** [comment]
Good afternoon Mr Bluca. Here's my snapshot of the log message sir: 
![log_message](https://user-images.githubusercontent.com/109356020/223736813-f630da8e-67f6-4c32-88d2-af65ab54c93b.PNG)

**Comment by 1awesomeJ** [comment]
I'd like to comment on the main function in the "src/ac-power/ac-power.c" module

The main function with a macro which calls a "run()" function. The run function calls a "parse_argv" function and passes the "argc" and "argv" variables of main to it. The "parse_argv" function checks that argc > 0 and that argv != NULL. The "parse_argv" function calls the "getopt_long" function which parses the command lines arguments and returns one of 3 integer values defined in a struct. The values are used in a switch case by "parse_argv" to determine the value to return to "run()". If the returned value is greater than 0, "run()" calls an "on_ac_power()" function which doesn't seem defined in the "src/ac-power/ac-power.c" module. It's probably defined in one of the included header files.

**Comment by bluca** [comment]
> I'd like to comment on the main function in the "src/ac-power/ac-power.c" module

Ok, you can proceed with the short tasks now

**Comment by 1awesomeJ** [comment]
Alright sir.

**Comment by 1awesomeJ** [comment]
Good evening Mr Bluca,
1. could you give us times when it's okay to call on you here, just so we don't do it at the wrong times.

2. About summarizing one of the documentation, Is the documentation on coding style appropriate for summarization?

Thank you

**Comment by bluca** [comment]
> Good evening Mr Bluca,
> 
> 1. could you give us times when it's okay to call on you here, just so we don't do it at the wrong times.

Anytime, I'll see it when I see it

> 2. About summarizing one of the documentation, Is the documentation on coding style appropriate for summarization?

You don't need to do that, you've done enough of the first block, you can move to the second block (short tasks)

**Comment by 1awesomeJ** [comment]
Alright sir. I'll start studying. I'll be glad to submit a PR by tomorrow.

**Comment by 1awesomeJ** [comment]
> Anytime, I'll see it when I see it

Awesome

**Comment by 1awesomeJ** [comment]
I'm currently trying to work on the issue: missing specifiers in user.conf #26596

The person said the %h specifier which should be for "user home dir" isn't available to unit files.

He said the list of available specifiers is in the file "~/.config/systemd/user.conf" and detailed in "systemd-system.conf(5)" 

I figured "opening up the files to see their contents would be a good start"

However, I tried locating the files on my machine, and even in the systemd/ repo I cloned, I couldn't find them.

Any help on what to do would be greatly appreciate sir.

**Comment by 1awesomeJ** [comment]
Oh... Okay sir.

All the 16 open issues have the yellow tag "please-submit-as-pr" How can I know which one to work on sir?

**Comment by bluca** [comment]
check if a PR is linked to the issue

**Comment by 1awesomeJ** [comment]
This one here: "[systemd-boot exits on secure boot error w/o trying next boot option #25548](https://github.com/systemd/systemd/issues/25548)"

It's about someone who has two kernel images on his machine, and would love for the second image to be checked by systemd-boot, if booting fails on the first. Currently systemd-boot just exits when it fails on the first image.

It looks interesting. It looks tough too. The fact that it's stayed open for more than 3 months also scares me.

I'm like "Is it that the experienced developers haven't fixed it since?, what chance does a beginner like me have?"

What would you say to that sir?

**Comment by 1awesomeJ** [comment]
Good morning Bluca.
I want to see if I can work on some unit tests today.

Reading through the internship project again, I realized that the overall booting experience is what we actually want to improve for users, by adding modern features. Based on that, I'll definitely be working on the systemd-boot issue of exiting without checking for other kernel images that may be present.


It's my first time contributing to such a large and far-reaching project.

Are there specific resources or documentations you could share with me to help flatten the learning curve somewhat?

I'll definitely see this through.

Thank you sir.

**Comment by 1awesomeJ** [comment]
err = image_start(image, entry);
                if (err != EFI_SUCCESS)
                        goto out;
The guy who raised the issue mentioned that he thinks image_start() is returning an error, which is making "go-to out" happen.

That code snippet is on line 2400+

I've never worked on code that's more than 200 lines.

But I want to ask,

Instead of having image_start(image, entry)
Could we have an images array instead:
Image_start(images[ ], entry)
So that we then do:
k= 0, 
While image[ k] {
err = image_start(images[k], entry);
                if (err != EFI_SUCCESS)
                        goto out;
               k++;
}

**Comment by bluca** [comment]
You could certainly try that approach, see the documents listed in the first post here, especially https://systemd.io/HACKING/ as you'll need to use mkosi to test systemd-boot

**Comment by 1awesomeJ** [comment]
Thank you so much sir. 
The prospect of my internship is one thing that is exciting.
The prospect of working on such a huge project as 'systemd' that is used by tens of millions of people globally is even more exciting. 
How about the privilege of getting mentored by one of the core maintainers of the Debian distro?
Please be patient with me sir. I'll commit myself more and more.

**Comment by 1awesomeJ** [comment]
Good evening Mr Bluca, I went through the code in https://github.com/systemd/systemd/tree/v252/src/boot/efi)/boot.c

I don't fully understand it yet, but in trying to implement my idea of an array of kernel images, I need to modify this line:
 err = image_start(image, entry);

I found the fuction image_start  (lines 2351 -2438)

But I didn't find the variables "image" and "entry".

Maybe they're in one of the included headers.

What would you advise sir?

**Comment by 1awesomeJ** [comment]
Good afternoon Mr Bluca,

I'm trying to work on the task of improving an existing unit test is src/test.

However I have some level of confusion.

I went through the the src/basic directory.
And I checked the argv-util.c module.

The function I would like to create tests for is:

"book argv_looks_like_help( int argc, char **argv)" function.

Is it the "test-argv-util.c" file inc src/test 
that I have to create my test in?

If yes?

How can I test locally?
Do I have to rebuild with meson?

I tried doing "gcc test-argv-util.c"
The compiler complained of a missing header file.

I'm sorry for having naive questions. I just want to break the initial hurdles of this process which is quite new to me.

**Comment by Agabaderrickjunior** [comment]
hello there am Derrick an outreachy applicantant new to this project

**Comment by 1awesomeJ** [comment]
> hello there am Derrick an outreachy applicantant new to this project

Welcome Derrick.. Lots of deep C programming to do.

**Comment by bluca** [comment]
> Good afternoon Mr Bluca,
> 
> I'm trying to work on the task of improving an existing unit test is src/test.
> 
> However I have some level of confusion.
> 
> I went through the the src/basic directory. And I checked the argv-util.c module.
> 
> The function I would like to create tests for is:
> 
> "book argv_looks_like_help( int argc, char **argv)" function.
> 
> Is it the "test-argv-util.c" file inc src/test that I have to create my test in?
> 
> If yes?
> 
> How can I test locally? Do I have to rebuild with meson?
> 
> I tried doing "gcc test-argv-util.c" The compiler complained of a missing header file.
> 
> I'm sorry for having naive questions. I just want to break the initial hurdles of this process which is quite new to me.

Yes, build with meson, and then you'll find the test file in the build/ directory, and you can just call it and it will run

**Comment by 1awesomeJ** [comment]
Awesome. I'm so glad I got this response sir.
Thank you so much.

**Comment by 1awesomeJ** [comment]
> > Good afternoon Mr Bluca,
> > I'm trying to work on the task of improving an existing unit test is src/test.
> > However I have some level of confusion.
> > I went through the the src/basic directory. And I checked the argv-util.c module.
> > The function I would like to create tests for is:
> > "book argv_looks_like_help( int argc, char **argv)" function.
> > Is it the "test-argv-util.c" file inc src/test that I have to create my test in?
> > If yes?
> > How can I test locally? Do I have to rebuild with meson?
> > I tried doing "gcc test-argv-util.c" The compiler complained of a missing header file.
> > I'm sorry for having naive questions. I just want to break the initial hurdles of this process which is quite new to me.
> 
> Yes, build with meson, and then you'll find the test file in the build/ directory, and you can just call it and it will run

This has really been helpful.
I've been able to move from being very confused to have much clarity.

The fact that I also saw specific lines that need tests on Coveralls has also helped.


I'm defining a function "test_argc_looks_like_help(char *argv[ ], int argc)"

inside the test-argv-utils.c file.

One of the compiler errors I got was "no prior function declaration"

1. In which header file do I declare my test function? "tests.h"?
2. In which directory am I likely to find the header file?
Thank you sir.

**Comment by bluca** [comment]
define it as `static`, no need for a header

**Comment by 1awesomeJ** [comment]
> define it as `static`, no need for a header

Wow! This is awesome.
I was actually wondering why some of the other functions were defined as static.

Thank you so much sir.

**Comment by 1awesomeJ** [comment]
Good afternoon Mr Bluca, My test shows now:
![Uploading test2.PNG…]()

**Comment by 1awesomeJ** [comment]
![test2](https://user-images.githubusercontent.com/109356020/224481422-9f22267c-1a8c-4045-b5a6-936950e5345a.PNG)

**Comment by 1awesomeJ** [comment]
Although I still have an issue:
I called my test function with the TEST macro that was used for all the other tests.

The test macro wouldn't work for my function while its prototype was "static void test_argv_looks(char **args, int argc)"
It only works when I use the prototype 
"static void test_argv_looks(char *args, int argc)"

I.e it only works with a single string and not an array of strings.

It complained of conflicting types in the definition of the macro in src/shared/tests.h

Is there a way around this sir?

I thought of defining another macro "TEST2" in src/shared/tests.h, 
but as much as possible, I'd like to avoid touching the tests.h

If everyone who wrote a unit test had to modify the header file, it would have been way too clumsy.

**Comment by bluca** [comment]
there's no need for a new macro - open a pull request with what you have, and we can have a look

**Comment by 1awesomeJ** [comment]
> there's no need for a new macro - open a pull request with what you have, and we can have a look

Okay sir.

I have just created a new pull request sir.

**Comment by 1awesomeJ** [comment]
![test3](https://user-images.githubusercontent.com/109356020/224491273-b56411a5-cfc8-40e1-b9f3-2fad8f5907cf.PNG)

**Comment by taniishkaaa** [comment]
Hi, I'm Tanishka, an Outreachy applicant. I would like to contribute to the systemd project. Looking forward to connecting and contributing!

**Comment by taniishkaaa** [comment]
How can I get started?

**Comment by bluca** [comment]
check the instructions at the top

**Comment by taniishkaaa** [comment]
Hello! 
I have successfully built the project in ubuntu. How can I proceed with the logging the message part?

**Comment by 1awesomeJ** [comment]
> Hello! I have successfully built the project in ubuntu. How can I proceed with the logging the message part?

You could try reading through some of the earlier messages.  You will choose any of the tools in src/ to add your log message to.
I am trying to navigate myself.

**Comment by 1awesomeJ** [comment]
@bluca,
I got your approval on my PR sir.
1. Will it be merged to the main branch?
2. Can I start attempting another short task? "add a new unit test to cover the 'shared' common code"
3. Should I wait till it's merged before I record it as a contribution on my Outreachy application?

**Comment by 1awesomeJ** [comment]
> @bluca, I got your approval on my PR sir.
> 
> 1. Will it be merged to the main branch?
> 2. Can I start attempting another short task? "add a new unit test to cover the 'shared' common code"
> 3. Should I wait till it's merged before I record it as a contribution on my Outreachy application?

Good morning Mr Bluca, I'm so grateful for all your help so far helping me have my first PR succesfully merged. Thank you sir.
The merge has answered my questions 1 and 3 here, so I'll just move to the task of adding a new unit test to cover the 'shared' common code.

I'll ask for help when I hit any blockers.

Thank you sir.

**Comment by 1awesomeJ** [comment]
Good morning Mr Bluca, I've gone through some of the files in src/shared/ and I'm thinking of creating test for the 
"devnode-acl.c" code.
Is that alright sir?
I'll be glad to open a PR on it, so I get to learn from you and the other senior members again sir.

**Comment by bluca** [comment]
Sure, go ahead

**Comment by 1awesomeJ** [comment]
Alright sir

**Comment by Agabaderrickjunior** [comment]
hello @1awesomeJ @bluca   which are the first issues some one an begin with for the first time

**Comment by Agabaderrickjunior** [comment]
> `root@Ahmed-Laptop:~/systemd/build/build# apt build-dep systemd Reading package lists... Done E: You must put some 'deb-src' URIs in your sources.list root@Ahmed-Laptop:~/systemd/build/build#`
> 
> It shows this now, what do I do next?

hey did some one get solution to this problem ...been facing the challenge and came back to viist the thread but got no solution 
@bluca @1awesomeJ @sidrahh12

**Comment by bluca** [comment]
> hello @1awesomeJ @bluca which are the first issues some one an begin with for the first time

Please see the top post, you need to start with "repeatable tasks".

**Comment by bluca** [comment]
> > `root@Ahmed-Laptop:~/systemd/build/build# apt build-dep systemd Reading package lists... Done E: You must put some 'deb-src' URIs in your sources.list root@Ahmed-Laptop:~/systemd/build/build#`
> > It shows this now, what do I do next?
> 
> hey did some one get solution to this problem ...been facing the challenge and came back to viist the thread but got no solution @bluca @1awesomeJ @sidrahh12

Please use the communication/support channels of your distribution for help on how to install packages

**Comment by 1awesomeJ** [comment]
> > `root@Ahmed-Laptop:~/systemd/build/build# apt build-dep systemd Reading package lists... Done E: You must put some 'deb-src' URIs in your sources.list root@Ahmed-Laptop:~/systemd/build/build#`
> > It shows this now, what do I do next?
> 
> hey did some one get solution to this problem ...been facing the challenge and came back to viist the thread but got no solution @bluca @1awesomeJ @sidrahh12

Normally I would google the entire line of the error message. What i remember is that there were about 5 dependency packages that were missing on my machine (starting with 'gperf' or so) I just installed them one after the other, and i was able to build the project successfully.

> On Ubuntu you can use `apt build-dep systemd` to install the required build dependencies

You could also refer to Mr Bluca's message here to install all the required dependencies.

**Comment by 1awesomeJ** [comment]
Good evening Mr Bluca, I just created a new PR on the short task  "update outdated executables/daemons to bus_call_method/bus_get_property"

https://github.com/systemd/systemd/pull/26819

**Comment by 1awesomeJ** [comment]
I've been having issues with my aws ubuntu instance since yesterday. I couldn't get it resolved.
I eventually chose to use a vagrant box.

I'm glad I was able to do a PR today.

Thank you to all of you mentors for always insisting on good quality code for PR, it's always awesome to get your reviews.

**Comment by 1awesomeJ** [comment]
Mr Bluca, I'm so sorry I delayed pushing my PR corrections. My vagrant box started some downloading some installation packages again.

Thankfully, I have now pushed the corrections. Yours and Mr DaanDeMeyer's.

**Comment by ZIHCO** [comment]
Hi Bluca, I am Chizoba ODINAKA. And I am interested in contributing to the Systemd project.

**Comment by bluca** [comment]
welcome, please see instructions at the top

**Comment by 1awesomeJ** [comment]
Good evening Mr Bluca
I just made a PR for the short task "update outdated executables/daemons to new-style sd-bus macros:"

https://github.com/systemd/systemd/pull/26840

**Comment by 1awesomeJ** [comment]
Good morning Mr Bluca,
having done a few short tasks, I would love for my next PR to be on one of the long tasks.
I would study over the next few days to be able to come up with a working solution.

If I have issues, I'll be sure to mention them to you here.

Thank you for your consistent support sir.

**Comment by bluca** [comment]
Sounds good, I'd recommend to start with adding a non-simple test case for some existing library/tool

**Comment by 1awesomeJ** [comment]
> Sounds good, I'd recommend to start with adding a non-simple test case for some existing library/tool

Alright sir.
Thank you so much.

**Comment by 1awesomeJ** [comment]
If I may ask sir what tool/library would you suggest I look at?

**Comment by bluca** [comment]
Check the coverage report, something that has partial coverage but not complete. Also something that you can run easily locally/in a VM to familiarize yourself with

**Comment by 1awesomeJ** [comment]
> Check the coverage report, something that has partial coverage but not complete. Also something that you can run easily locally/in a VM to familiarize yourself with

Thank you so much sir.

I'll drop a list of the tools I find later today.

**Comment by taniishkaaa** [comment]
![f7171c39-a04e-4e3d-80e7-0a1df13ca0e4](https://user-images.githubusercontent.com/109246904/225689186-a28551a7-be6f-472b-83a0-bc6a2c748d98.jpg)
Hi @bluca here is my snapshot of one of the repeatable tasks. Is this okay?

**Comment by bluca** [comment]
looks good

**Comment by taniishkaaa** [comment]
and here is the summary of the document "Contributing" in a paragraph:
The guidelines for contributing to the systemd project on GitHub include using GitHub Issues for tracking bugs and feature requests, specifying the systemd version and distribution used, following coding style when contributing code and testing changes before submitting Pull Requests (PRs), and ensuring compatibility with LGPL-2.1-or-later license if copying existing code. Backward compatibility and external dependencies are considered on a case-by-case basis, and new features can require bleeding edge external dependencies but should have runtime and build-time fallbacks. Existing dependencies should not be bumped without strong reasons, and kernel requirements should also be applied with strong reasons required for breaking compatibility for existing functionality. The maintainers may not be able to respond immediately to issues or PRs.

**Comment by bluca** [comment]
You can proceed to the short tasks phase

**Comment by taniishkaaa** [comment]
can we proceed with any of the short tasks listed above and how many of them are we supposed to implement?

**Comment by taniishkaaa** [comment]
okay! thankyou

**Comment by 1awesomeJ** [comment]
Good evening Mr Bluca, 
I'm considering writing tests for the "reboot-to-firmware" tool in "src/boot/bootctl-reboot-to-firmware.c". 

I hope to get lots of guidance sir.

**Comment by bluca** [comment]
There's no test framework in place for the bootloader stuff, so I'd recommend to avoid that

**Comment by 1awesomeJ** [comment]
> There's no test framework in place for the bootloader stuff, so I'd recommend to avoid that

Ohh..Okay sir

**Comment by 1awesomeJ** [comment]
I'll search for another tool

**Comment by 1awesomeJ** [comment]
Good morning Mr Bluca, thank you again for helping avoid working on the wrong test yesterday.


I found another candidate: The resize-fs tool in 
src/shared/resize-fs.c

Is it a good one sir?

**Comment by bluca** [comment]
sure

**Comment by 1awesomeJ** [comment]
> sure

Alright sir. Thank you.

**Comment by 1awesomeJ** [comment]
Good afternoon Mr Bluca,
I have opened a PR with what i have on the test.
https://github.com/systemd/systemd/pull/26868

**Comment by 1awesomeJ** [comment]
About the TO-DO items, can we be sure that any item we find on it is still open sir?

I'd like to study for a few of them to see which I can implement.

**Comment by 1awesomeJ** [comment]
Good morning Mr Bluca.
My PR for the task "add a new unit test to cover the 'shared' common code" is still waiting for your review.

I want to work on the long task: "fix one of the TODO items" 

I saw in the list where it says to " use secure_getenv() instead of getenv() where appropriate"
Is that an appropriate one?


I'm also looking at the task for writing integration test. Would it likely require testing with mkosi?

**Comment by taniishkaaa** [comment]
Hey @bluca! For my short tasks, I would like to work on #25293

**Comment by bluca** [comment]
> Hey @bluca! For my short tasks, I would like to work on #25293

Sure, there's no need to ask, just open PRs

**Comment by bluca** [comment]
> Good morning Mr Bluca. My PR for the task "add a new unit test to cover the 'shared' common code" is still waiting for your review.
> 
> I want to work on the long task: "fix one of the TODO items"
> 
> I saw in the list where it says to " use secure_getenv() instead of getenv() where appropriate" Is that an appropriate one?

Sure

> I'm also looking at the task for writing integration test. Would it likely require testing with mkosi?

No that uses some local scripts, see the readme in test/

**Comment by 1awesomeJ** [comment]
Good evening Mr Bluca,
I opened a PR for the task "add a new unit test to cover the 'shared' common code"
https://github.com/systemd/systemd/pull/26891

**Comment by 1awesomeJ** [comment]
Good morning Mr Bluca, my PR https://github.com/systemd/systemd/pull/26891 is still open.
It's failing checks on Github, but i tested it locally, and made sure it worked before pushing.
When I had thesame issue on Sunday 12th March, @DaanDeMeyer mentioned that the failures were unrelated to my code.
I want you to please help review sir.


Now about the integration test, I'm going through the README.testsuite, I want to ask, will our work likely involve modifying the test/test-functions file?

**Comment by bluca** [comment]
> Now about the integration test, I'm going through the README.testsuite, I want to ask, will our work likely involve modifying the test/test-functions file?

Depends on what the test does, not necessarily if everything is already supported

**Comment by 1awesomeJ** [comment]
> > Now about the integration test, I'm going through the README.testsuite, I want to ask, will our work likely involve modifying the test/test-functions file?
> 
> Depends on what the test does, not necessarily if everything is already supported

Oh...Thanks sir.

I thought you were away at some conference.

**Comment by taniishkaaa** [comment]
@bluca I've created a PR for the issue (#22793) here #26905

**Comment by 1awesomeJ** [comment]
Good morning Mr Bluca,
I went through Coveralls to check the impact of my last unit test on coverage. The coverage is currently at 97.92% for the module. I'd like to make it 100%.
Is there a way I could do this without opening another PR?
It's just one line of code sir.

**Comment by bluca** [comment]
No, you need to open a new PR

**Comment by 1awesomeJ** [comment]
> No, you need to open a new PR

Alright. I'll do that after I submit one for the integration test sir.

**Comment by 1awesomeJ** [comment]
Good morning Mr Bluca,
I'm reading that the name of the dbus interface we use is "org.freedesktop/systemd1"?
How accurate is that?

**Comment by bluca** [comment]
yes that is all correct

**Comment by 1awesomeJ** [comment]
Good morning Mr Bluca,
1. Our task on the integration tests, is it to add a new test e.g "TEST-80-LOGIN"?
2. When I ran the existing tests locally 74 out of 79 failed. The failure from "setfacl". It was "setfacl:command not found" Do we need to install any dependencies to make it run.
So that we'll be able to test locally before pushing.

Thank you sir.

**Comment by 1awesomeJ** [comment]
3. Is there a hint on how we could know which tools are not yet covered. I've been trying to find sir.

**Comment by 1awesomeJ** [comment]
> 1 - yes but there's already a logind test 2 - you'll need to install some extra packages, this list here should be everything that's needed https://salsa.debian.org/systemd-team/systemd/-/blob/upstream-ci/debian/tests/control#L121 3 - check the coveralls report

Thank you so so much sir.

**Comment by 1awesomeJ** [comment]
Good evening Mr Bluca,
Please I want to ask if the "usermod(nologin)"
tool would be appropriate for me to work on for the integration test.

**Comment by 1awesomeJ** [comment]
I mean the module/src/user-sessions/user-sessions.c

Could I write an integration test for its behavior?

**Comment by bluca** [comment]
it's already well covered and it's tiny, so probably not the most significant one

**Comment by 1awesomeJ** [comment]
> it's already well covered and it's tiny, so probably not the most significant one

Thank you so much sir.
I'll look for another one.

**Comment by 1awesomeJ** [comment]
Good evening Mr Bluca,
I saw somewhere that mkosi is just for creating OS images that can be booted, while nspawn creates an entire containerized VM.

1. Is that true?
2. Is it only in tests and issues that involve sd-boot (booting) that mkosi is relevant?
3. When I run "test/run-integration-tests.sh" on my machine, I know some tests are set to use nspawn, others qemu, others either? Is it the nspawn or qemu that executes those tests or my local machine.

**Comment by 1awesomeJ** [comment]
Good morning Mr Bluca, I went on Coveralls again, and I saw test coverage for openssl tool at 36.96%.
That's /src/shared/openssl-util.c
would that be a good one to write the integration test for?

**Comment by 1awesomeJ** [comment]
I'm also looking at the "/src/basic/random-util.c" which currently has coverage of 33.86%.
I don't fully understand it yet, but what I've gathered so far is with the "systemd-cryptenroll" command, i could have a test script that will cover the "/src/basic/random-util.c"(33.86%) as well as "/src/shared/recovery-key.c" (55.93%) and  "/src/crpytenroll/cryptenroll.c" (51.23%).

should i make the attempt sir?

sorry for questions that sometimes get too many.

**Comment by 1awesomeJ** [comment]
I noticed most modules with "-util" suffix don't have direct user-facing commands but are instead used by other tools.

**Comment by bluca** [comment]
> Good evening Mr Bluca, I saw somewhere that mkosi is just for creating OS images that can be booted, while nspawn creates an entire containerized VM.
> 
> 1. Is that true?
> 2. Is it only in tests and issues that involve sd-boot (booting) that mkosi is relevant?
> 3. When I run "test/run-integration-tests.sh" on my machine, I know some tests are set to use nspawn, others qemu, others either? Is it the nspawn or qemu that executes those tests or my local machine.

1) no, check the manpages
2) not only
3) both, depending on the test, some need a virtual machine, some don't

> Good morning Mr Bluca, I went on Coveralls again, and I saw test coverage for openssl tool at 36.96%. That's /src/shared/openssl-util.c would that be a good one to write the integration test for?

Everything in shared/ is a library, not a tool, so a unit tests under src/test/ is more appropriate

> I'm also looking at the "/src/basic/random-util.c" which currently has coverage of 33.86%. I don't fully understand it yet, but what I've gathered so far is with the "systemd-cryptenroll" command, i could have a test script that will cover the "/src/basic/random-util.c"(33.86%) as well as "/src/shared/recovery-key.c" (55.93%) and "/src/crpytenroll/cryptenroll.c" (51.23%).

cryptenroll is already partially covered in test/units/testsuite-70.sh, so you can add to that to increase coverage

**Comment by 1awesomeJ** [comment]
Wow. You're a life saver sir.
Been on this for 3 days straight.

"Everything in shared is a library not a tool"
I'll remember that strongly.

Let me work on the test/units/testsuite-70.sh sir.
It's such a relief

**Comment by taniishkaaa** [comment]
Hey @bluca I opened a PR for the issue (#12914) here #26992

**Comment by bluca** [comment]
> Hey @bluca I opened a PR for the issue (#12914) here #26992

That's fine. For the next contribution you should look at submitting some code changes

**Comment by taniishkaaa** [comment]
Do you mean like improving the unit tests or working on different labeled issues?

**Comment by bluca** [comment]
Anything you like, as long as there are code changes involved - fixing some hwdb configuration is fine to get started, but it is not sufficient by itself, as it's just configuration

**Comment by taniishkaaa** [comment]
Noted, thanks.

**Comment by 1awesomeJ** [comment]
Good afternoon Mr Bluca,
I opened a PR for the integration test. I added tests to testsuite-70.sh to cover the cryptenroll.c module:

https://github.com/systemd/systemd/pull/27007

1. Most of the lines not yet covered on Coveralls are inside the parse_argv() function.
There's about 300 lines of the function(inluding blank lines) that aren't yet covered.
I started with the first 100 lines and I'd like to get your feedback while I work on the remaining lines.

2. I aslo added a line to the test-compare-operator.c unit test I wrote in my earlier PR in expectation that the module
will hit a 100% coverage.

some checks are failing on Github.

**Comment by 1awesomeJ** [comment]
mkosi/ci tests

**Comment by 1awesomeJ** [comment]
@bluca, 
while trying to break the unit test into a separate commit, i had issues with the git. So I closed the PR and reopend another one here: 
https://github.com/systemd/systemd/pull/27020

**Comment by 1awesomeJ** [comment]
I was also unable to comment on the old PR. I tried severally to respond to the reviews but I kept getting "you can't comment at this time"

**Comment by taniishkaaa** [comment]
Hi @bluca I would like to work on issue #27028.
My approach to the fix is to implement a buffer system to hold requests until workers become available in userwork.c
Do you think that would be an appropriate solution?

**Comment by bluca** [comment]
I'd recommend to get started with something easier, like improving an existing unit test

**Comment by 1awesomeJ** [comment]
Good evening sir.
My PR https://github.com/systemd/systemd/pull/27020 is ready for your review.

**Comment by 1awesomeJ** [comment]
For my long task "fix an issue labelled RFE",
I would love to work on https://github.com/systemd/systemd/issues/26697
Can I be assigned to it sir?

**Comment by bluca** [comment]
sure

**Comment by 1awesomeJ** [comment]
> sure

Alright sir.

**Comment by Wendy-wyt** [comment]
![image](https://user-images.githubusercontent.com/78782949/228417448-a8bb1d0b-b083-4d6f-a38d-d8f2efd554bf.png)

**Comment by bluca** [comment]
Looks good, you can proceed

**Comment by bluca** [comment]
looks good, you can proceed to a short task

**Comment by Wendy-wyt** [comment]
Thank you! Are short tasks repeatable? If no, which short tasks can I work on?

**Comment by bluca** [comment]
They are not, but there's still plenty of open ones in the list, so you can pick any of them

**Comment by 1awesomeJ** [comment]
Good morning Mr Bluca, I'm not too familiar with XML. So I'm trying to render the current manpage I'm working on with my web browser so I can see the updates I make. The browser complains of "encoding errors", doesn't fully render the content, and doesn't display the tables.
what do you advise sir?

**Comment by 1awesomeJ** [comment]
Good morning Mr Bluca, I opened a PR for my long task 'fix one of the issues labelled as "RFE"'.
https://github.com/systemd/systemd/pull/27065
kindly help review sir.

**Comment by 1awesomeJ** [comment]
> > Good morning Mr Bluca, I'm not too familiar with XML. So I'm trying to render the current manpage I'm working on with my web browser so I can see the updates I make. The browser complains of "encoding errors", doesn't fully render the content, and doesn't display the tables. what do you advise sir?
> 
> Enable building manpages with `meson configure build -Dman=true` and then rebuild, and you'll be able to read the generated manpage with `man ./build/man/some-man-page`

Wow, thank you so much sir.

**Comment by 1awesomeJ** [comment]
Good evening sir, 
Here's one of the items on the TODO list:

* revisit default PCR bindings in cryptenroll and systemd-creds. Currently they
  use PCR 7 which should contain secureboot state db/dbx. Which sounded like a
  safe bet, given that it should change only on policy changes, and not
  software updates. But that's wrong. Recent fwupd (rightfully) contains code
  for updating the dbx denylist. This means even without any active policy
  change PCR 7 might change. Hence, better idea might be in systemd-creds to
  default to PCR 15 at least if sd-stub is used (i.e. bind to system identity),
  and in cryptsetup simply the empty list? Also, PCR 14 almost certainly should
  be included as much as PCR 7 (as it contains shim's policy, which is
  certainly as relevant as PCR 7 on many systems)

**Comment by 1awesomeJ** [comment]
I'm considering it for my long task "fix one of the TODO items"
Is it a good one?

**Comment by bluca** [comment]
That's more of a policy change, there won't be much coding to do, so better to look at something else

**Comment by 1awesomeJ** [comment]
> That's more of a policy change, there won't be much coding to do, so better to look at something else

Oh...
I'll go fish for a better one then.

**Comment by 1awesomeJ** [comment]
Good morning Mr Bluca,
I made all the fixes on my RFE PR, but the integration test caused a conflict.
In trying to resolve, I accidentally did a reset on the branch that took out my commits and closed the PR.

I have reopened it here:
https://github.com/systemd/systemd/pull/27091

**Comment by 1awesomeJ** [comment]
Good morning Mr Bluca, for my long task: "fix one of the TODO items"
I'm considering either or both of these:
* get rid of prefix_roota() and similar, only use chase() and related
  calls instead.

* get rid of basename() and replace by path_extract_filename()

First, are they good enough?, even if I only manage to implement one and not both?

I plan on submitting my final application tomorrow, as the deadline is in two days.

What do you advise sir? about the TO DO items?

**Comment by bluca** [comment]
Yeah that's fine, go ahead with either or both

**Comment by 1awesomeJ** [comment]
> Yeah that's fine, go ahead with either or both

Alright sir.

**Comment by taniishkaaa** [comment]
Hey @bluca! 
Can you help me understand the step-wise timeline for each of the tasks to be performed during the internship period for my final application?

**Comment by bluca** [comment]
The tasks are described at https://www.outreachy.org/outreachy-may-2023-internship-cohort/communities/systemd/#improve-user-experience-on-early-system-boot-failu and should be all doable within the internship timeframe

**Comment by taniishkaaa** [comment]
> The tasks are described at https://www.outreachy.org/outreachy-may-2023-internship-cohort/communities/systemd/#improve-user-experience-on-early-system-boot-failu and should be all doable within the internship timeframe

Do we have to describe the time commitments for each task or can we just list them in the application?
Also, I've tried writing a unit-test for a function in the env-util.c file. Do I simply open a PR for that or would you like to approve the code first?

**Comment by bluca** [comment]
Just list them. And just open a PR.

**Comment by taniishkaaa** [comment]
Thanks so much. I opened the PR here #27100.

**Comment by 1awesomeJ** [comment]
Good evening sir. I opened a PR for the TODO item task:
https://github.com/systemd/systemd/pull/27102

I hope you find time to help review. It's failing some tests.

**Comment by bluca** [comment]
The application period is now over, closing, thanks for all the contributions
